monai/networks/nets/autoencoderkl.py

```diff
@@ -320,6 +321,38 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         return x + residual
 
 
+class _AttentionBlockNew(nn.Module):
+    def __init__(self, spatial_dims, num_channels, num_head_channels):
+        super().__init__()
+        from monai.networks.blocks.selfattention import SABlock
+
+        self.spatial_dims = spatial_dims
+        self.norm = nn.GroupNorm(num_groups=32, num_channels=num_head_channels, eps=1e-6, affine=True)
+        self.attn = SABlock(hidden_size=num_channels, num_heads=1, qkv_bias=True)
+
+    def forward(self, x):
+        residual = x
+        # SAblck expects input in the form of B x C x (s_dim_1 * ... * s_dim_n)
+        from einops.layers.torch import Rearrange
+
+        if self.spatial_dims == 2:
+            h, w = x.shape[2], x.shape[3]
+            rearrange_input = Rearrange("b c h w -> b (h w) c")
+            rearrange_output = Rearrange("b (h w) c -> b c h w", h=h, w=w)
+        if self.spatial_dims == 3:
+            h, w, d = x.shape[2], x.shape[3], x.shape[4]
+            rearrange_input = Rearrange("b c h w d -> b (h w d) c")
+            rearrange_output = Rearrange("b (h w d) c -> b c h w d", h=h, w=w, d=d)
+
+        x = self.norm(x)
+        x = rearrange_input(x)  # B x (s_dim_1 * ... * s_dim_n) x C
+
+        x = self.attn(x)
+        x = rearrange_output(x)
+        x = x + residual
+        return x
```

Comment:
```suggestion
class _AttentionBlockNew(SkipConnection):
    def __init__(self, spatial_dims, num_channels, num_head_channels):
        from monai.networks.blocks.selfattention import SABlock
        from einops.layers.torch import Rearrange

        _,_,h,w,*d=x.shape
        if not d:
            rearrange_input = Rearrange("b c h w -> b (h w) c")
            rearrange_output = Rearrange("b (h w) c -> b c h w", h=h, w=w)
        else:
            rearrange_input = Rearrange("b c h w d -> b (h w d) c")
            rearrange_output = Rearrange("b (h w d) c -> b c h w d", h=h, w=w, d=d[0])
                
        norm = nn.GroupNorm(num_groups=32, num_channels=num_head_channels, eps=1e-6, affine=True)
        attn = SABlock(hidden_size=num_channels, num_heads=1, qkv_bias=True)
        super().__init__(Sequential(norm,rearrange_input,attn,rearrange_output),mode="add")
```

Reply:
on further thought, this depends on x, which isn't available at init. this means we will need to put this in the forward, so i don't think we save much code inheriting the skip connection here

Reply:
Ah yes, we'd have to define our own block to use `Rearrange` for a given input shape. If we do that in multiple places this would work but if not then yeah we leave this as it is.